### DISCUSSION TITLE
Library for setting up RTC interrupts?

### DISCUSSION THREAD

**Discussion by ChocolateLoverRaj** [phil-opp/blog_os#1364] [Questions] [answered]
I found [`x86_rtc`](https://crates.io/crates/x86_rtc) which is good for getting the current time. Is there a library for setting up the interrupts (I know it's not hard, but still)?

I also created an issue at https://github.com/arceos-org/x86_rtc/issues/2

**Comment by tpdenk** [comment]
It's your OS, so you should probably roll your own solution. Keep in mind that interrupts should be fast, shouldn't lock and shouldn't allocate (all within reason of course). I tend to try to avoid interrupts because of those and more reasons.

If you want to have dynamic interrupts (i.e. be able to register certain tasks to be called upon interrupts), you should use a lock-free, allocation free collection. Intrusive linked lists come to mind. You can either implement one on your own (it's *really* challenging), or use [cordyceps' list](https://docs.rs/cordyceps/latest/cordyceps/struct.List.html).

Otherwise, depending on what you mean by "setting up the interrupts", [x2apic](https://github.com/kwzhao/x2apic-rs) comes to mind. Note that this is not a library to set up you IDT and stuff, but it allows you to configure your IOAPIC, which essentially routes hardware interrupts to cores.

---

If you want to have timekeeping, I'd recommend the following:
* set up HPET (no interrupts no nothing, just enable it)
* get an initial time from CMOS (no interrupts again, just read it as it's described on the OsDev wiki)
* have a "boot time" as a combination of CMOS time + HPET ticks, let's call this `REF`
* this is great for a lot of things in your os - you can implement an `Instant`, you have a working `Duration` type, you can measure, display wall time etc.
* once you have network access, get a timestamp with NTP
* recalculate `REF` with the NTP time + HPET offset (you probably want to stay monotonic)
* refresh `REF` every so often (staying monotonic)

---

* for timer interrupts, use a LAPIC, that's *really* easy to set up with interrupts, it's described in the tutorial also
* use LAPIC interrupts + HPET ticks for your preemptive scheduler

<sub>*(Some information in this answer is technically inaccurate, but I think it's sufficient to build an initial understanding. If you have questions, feel free to ask. If something in this answer contradicts OsDev or some other source, OsDev is probably right. Most design decisions are yours to make, and you can't really go wrong in a hobby os - `rustc` prevents the major foot guns.)*</sub>

**Reply by ChocolateLoverRaj** [reply]
I set up RTC interrupts cuz I felt like it and it was fun. Thanks for the guide. I will use LAPIC and HPET for more proper time. For now, RTC + RTC interrupt should be enough to show a clock with seconds with the current time.

**Comment by ChocolateLoverRaj** [comment, marked as answer]
I did not find a library so I added functions to https://github.com/arceos-org/x86_rtc/pull/3
